### PR TITLE
Merge of OGS XML Fragment Generator

### PR DESCRIPTION
**Pull request by bernardkwok**
Update: #445, #480 

- Core fixes (Niklas and myself):
-- Always have the output name to "out". Other names appear to fail.
-- Fix the ordering of Pw to be first (required in vertex input format)
-- Fix for some type castings in noise implementation
-- Emit LightingData structure
-- Make samplers not have a "uniform" prefix as they are function arguments which cannot have this qualifier (just override variable output for filename inputs).
-- Update tests.

- Generation integration
-- Replace prototype "wrapper" class with OGS XML Fragment generator which used GLSL Fragment generator.
-- Note that all '\r' character have to be removed from the buffer otherwise OGS removes all characters after this -- and thus removes ";" and "\n". This appears to be an OGS defect
-- Map from "Sampler" back to non sample post fix name for filename as this is what we bind to even though shading code uses the sampler.

- Command integration:
-- Make -x (explicit XML file) an option. By default logic should be to use dynamic wrapper generation. Explicit generation kept as work is still in progress so can compare generation vs hand-crafted.

- Binding integration
-- Update binding to match newly generated parameter naming conventions.

### CODE REVIEW

File: libraries/stdlib/genglsl/lib/mx_noise.glsl [hunk 1]
```diff
@@ -172,8 +172,8 @@ uint mx_hash_int(int x, int y)
 uint mx_hash_int(int x, int y, int z)
 {
     uint a, b, c;
-    uint len = 3;
-    a = b = c = 0xdeadbeef + (len << 2) + 13;
+    uint len = 3u;
+    a = b = c = uint(0xdeadbeef) + (len << 2u) + 13u;
```
Comment (bernardkwok) [inline libraries/stdlib/genglsl/lib/mx_noise.glsl]: Maximum uint value is 0xffffffff so 0xdeadbeef lies within the range and won't change behaviour.

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp [hunk 2]
```diff
@@ -34,60 +39,62 @@ MaterialXData::~MaterialXData()
...
 {
-    return _xmlFragmentWrapper.getPathOutputMap();
+    return _pathInputMap;
 }
 
 void MaterialXData::createDocument(const std::string& materialXDocumentPath)
 {
     // Create document
-    _document = MaterialX::createDocument();
+    _document = mx::createDocument();
```
Comment (bernardkwok) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: @ppenenko as code is touched will move to using mx as a namespace shortcut for code consistency with other mtlx code.

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp [hunk 3]
```diff
@@ -96,64 +103,124 @@ void MaterialXData::generateXML()
...
+        _generator.generate(_fragmentName, shader.get(), nullptr, stream);
+        _fragmentWrapper = stream.str();
+        if (_fragmentWrapper.empty())
+        {
+            return;
+        }
+        else
+        {
+            // Strip out any '\r' characters.
+            _fragmentWrapper.erase(std::remove(_fragmentWrapper.begin(), _fragmentWrapper.end(), '\r'), _fragmentWrapper.end());
```
Comment (bernardkwok) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: '\r' stripper.

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp [hunk 4]
```diff
@@ -96,64 +103,124 @@ void MaterialXData::generateXML()
...
+                const mx::ShaderPort* port = uniforms[i];
+                const std::string& path = port->getPath();
+                if (!path.empty())
                 {
-                    throw MaterialX::Exception("Failed to add OGS shader fragment from file.");
+                    std::string name = port->getVariable();
+                    if (port->getType()->getSemantic() == mx::TypeDesc::SEMANTIC_FILENAME)
+                    {
+                        // Strip out the "sampler" post-fix to get the texture name.
+                        size_t pos = name.find(mx::OgsXmlGenerator::SAMPLER_SUFFIX);
```
Comment (bernardkwok) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: We generate <filename>Sampler in the code so the variable has this name. We want back the "root" which is the name with "Sampler" stripped away as that is the Shader instance parameter name we're interested in.

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp [hunk 5]
```diff
@@ -96,64 +103,124 @@ void MaterialXData::generateXML()
...
+        return;
+    }
+
+    const bool fragmentExists = fragmentManager->hasFragment(fragmentName.c_str());
+    if (!fragmentExists)
+    {
+        MString fragmentNameM;
+
+        // Allow for an explicit XML file to be specified.
+        if (!ogsXmlPath.empty())
```
Comment (bernardkwok) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: Allow xml file override.

File: source/MaterialXContrib/MaterialXNode/MaterialXNode.cpp [hunk 6]
```diff
@@ -89,28 +91,20 @@ MStatus MaterialXNode::initialize()
...
-                CHECK_MSTATUS(nAttr.setHidden(false));
-
-                mdgModifier.addAttribute(thisMObject(), _outAttr);
-                //		CHECK_MSTATUS(addAttribute(_outAttr));
-            }
-        }
-	}
+    if (materialXData && materialXData->isValidOutput())
+    {
+        const MString outputName(MaterialX::OgsXmlGenerator::OUTPUT_NAME.c_str());
```
Comment (bernardkwok) [inline source/MaterialXContrib/MaterialXNode/MaterialXNode.cpp]: We hard-code the output name now so we don't need to worry about name changes. We always promote to float3 for type consistency, but will need to look at float4 output afterwards so we can get transparency working.

File: source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.h [hunk 7]
```diff
@@ -1,5 +1,5 @@
-#ifndef MaterialXShadingNodeImpl_H
-#define MaterialXShadingNodeImpl_H
+#ifndef MATERIALX_SHADINGNODE_IMPL_H
```
Comment (bernardkwok) [inline source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.h]: Minor: include naming convention change.

File: source/MaterialXGenOgsXml/OgsXmlGenerator.cpp [hunk 8]
```diff
@@ -137,6 +137,25 @@ namespace
         for (size_t i = 0; i < block.size(); ++i)
         {
             const ShaderPort* p = block[i];
+            if (p->getName() == "Pw")
```
Comment (bernardkwok) [inline source/MaterialXGenOgsXml/OgsXmlGenerator.cpp]: @ppenenko : unsure if documented but position *must* come first or the vertex/pixel structures will map to the wrong place as the vertex shader structure hard-codes Pm as the first struct member :(

File: source/MaterialXGenOgsXml/OgsXmlGenerator.cpp [hunk 9]
```diff
@@ -180,13 +199,14 @@ namespace
...
 
+const string OgsXmlGenerator::OUTPUT_NAME = "out";
 const string OgsXmlGenerator::SAMPLER_SUFFIX = "Sampler";
 
 OgsXmlGenerator::OgsXmlGenerator()
 {
 }
 
-void OgsXmlGenerator::generate(const Shader* glsl, const Shader* hlsl, std::ostream& stream)
+void OgsXmlGenerator::generate(const std::string& shaderName, const Shader* glsl, const Shader* hlsl, std::ostream& stream)
```
Comment (bernardkwok) [inline source/MaterialXGenOgsXml/OgsXmlGenerator.cpp]: We generate with a supplied name to ensure that the XML fragment name is set to a desired value vs (shader->getName()).

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp [hunk 10]
```diff
@@ -96,64 +103,124 @@ void MaterialXData::generateXML()
...
+        else
+        {
+            // Strip out any '\r' characters.
+            _fragmentWrapper.erase(std::remove(_fragmentWrapper.begin(), _fragmentWrapper.end(), '\r'), _fragmentWrapper.end());
+        }
+
+        // Extract out the input fragment parameter names along with their
+        // associated Element paths to allow for value binding.
+        const mx::ShaderStage& ps = shader->getStage(mx::Stage::PIXEL);
+        for (auto uniformsIt : ps.getUniformBlocks())
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: "auto" in this case resolves to the value type, not a reference, therefore each pair is copied, including a std::string. Should be "const auto&" instead. This antipattern is widespread in our code unfortunately.

Comment (ppenenko) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]:
```suggestion
        for (const auto& uniformsIt : ps.getUniformBlocks())
```

Comment (ppenenko) [reply source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: Also, the name of the variable is misleading. It's not an STL iterator and we're iterating over an STL container.

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp [hunk 11]
```diff
@@ -34,60 +39,62 @@ MaterialXData::~MaterialXData()
...
+}
+
+void MaterialXData::clearXML()
+{
+    _pathInputMap.clear();
+    _fragmentName.clear();
+    _fragmentWrapper.clear();
 }
 
 void MaterialXData::generateXML()
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: Should eventually become "generateXml" for consistency.

File: source/MaterialXContrib/MaterialXNode/MaterialXData.h [hunk 12]
```diff
@@ -52,40 +55,47 @@ struct MaterialXData
...
     /// as opposed to a texture grraph.
     bool elementIsAShader() const;
 
   protected:
     /// Returns whether the element is renderable
     bool isRenderable();
+
     void createDocument(const std::string& materialXDocument);
 
+    void clearXML();
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.h]: "clearXml"

File: source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp [hunk 13]
```diff
@@ -81,18 +94,18 @@ MStatus bindFileTexture(
             MHWRender::MTextureManager* textureManager = renderer->getTextureManager();
             if (textureManager)
             {
-                VP2TextureUniquePtr texture(
-                    textureManager->acquireTexture(imagePath.asString().c_str(), MaterialX::EMPTY_STRING.c_str())
-                );
-
+                MHWRender::MTexture* texture = textureManager->acquireTexture(imagePath.asString().c_str(), MaterialX::EMPTY_STRING.c_str());
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp]: Any reason for switching back from unique_ptr?

File: source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp [hunk 14]
```diff
@@ -102,7 +115,7 @@ MStatus bindFileTexture(
     // the MaterialX sampler state.
     const std::string SAMPLE_PREFIX_STRING("Sampler");
     std::string samplerParameterName(parameterName + SAMPLE_PREFIX_STRING);
-    Vp2SamplerUniquePtr samplerState{ MHWRender::MStateManager::acquireSamplerState(samplerDescription) };
+    const MSamplerState* samplerState = MHWRender::MStateManager::acquireSamplerState(samplerDescription);
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp]: Again, why not keep the unique_ptr? The sampler is leaked now.

File: source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp [hunk 15]
```diff
@@ -212,134 +209,121 @@ void MaterialXShadingNodeImpl<BASE>::updateDG()
...
-    const auto* const node = dynamic_cast<MaterialXNode*>(depNode.userNode());
+    MaterialXNode* node = dynamic_cast<MaterialXNode*>(depNode.userNode());
     if (!node)
     {
         return;
     }
 
-    // Get the parameter list for checking against.
-    MStringArray parameterList;    
+    // Get the parameter list fo checking against.
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp]: It looks like you've resolved by overwriting my changes to this file.

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp [hunk 16]
```diff
@@ -34,60 +39,62 @@ MaterialXData::~MaterialXData()
...
 {
+    // Reset cached data
+    clearXML();
```
Comment (feldstj) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: guess this should be clearXml then :)

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp [hunk 17]
```diff
@@ -96,64 +103,124 @@ void MaterialXData::generateXML()
...
+{
+    MHWRender::MRenderer* theRenderer = MHWRender::MRenderer::theRenderer();
+    MHWRender::MFragmentManager* fragmentManager = theRenderer ? theRenderer->getFragmentManager() : nullptr;
+    if (!fragmentManager)
+    {
+        return;
+    }
+        
+    // TODO: This should not be hard-coded
+    std::string dumpPath("d:/work/shader_dump/");
```
Comment (feldstj) [inline source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: Should we remove this dumpPath stuff? It points to a directory on your machine and it was never really that useful for me anyways even when I specified a valid path.

File: source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp [hunk 18]
```diff
@@ -8,6 +8,21 @@
 #include <maya/MPxShadingNodeOverride.h>
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXNode/MaterialXShadingNodeImpl.cpp]: Can I ask you to re-merge this file, transferring important changes manually? (see comments below)

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp (hunk 10 above)
Comment (bernardkwok) [reply source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: Thanks for catching this! I've scanned and patched up other places with issue.

File: source/MaterialXContrib/MaterialXNode/MaterialXData.cpp (hunk 17 above)
Comment (bernardkwok) [reply source/MaterialXContrib/MaterialXNode/MaterialXData.cpp]: It will need to be fixed up but it's the only way I can debug for now.

**Pull request merged**